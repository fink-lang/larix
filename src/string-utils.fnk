unindent_text = fn text, ind:
  [first, ...rest] = text.split rx/\n/g

  [...lines] = pipe rest:
    map line:
      line.slice ind

  [first, ...lines].join '\n'
