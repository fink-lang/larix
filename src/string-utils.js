import {curr_loc, next_loc} from '@fink/prattler';


export const unindent_text = (text, ind)=> (
  text
    .split(/\n/g)
    .map((line, idx)=> line.slice(idx > 0 ? ind : 0))
    .join('\n')
);


export const strip_block = (strings, ...parts)=> {
  const [, ...lines] = String.raw({raw: strings}, ...parts).split('\n');
  const ind = lines[0].search(/[^ ]/);
  const str = lines.map((lne)=> lne.slice(ind)).join('\n');
  return str;
};

