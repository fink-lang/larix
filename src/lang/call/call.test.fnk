{describe, it, eq, to_throw, snapshot} = import '../../testing/jest'

{parse_expr} = import '../..'


describe:: 'call()', fn:

  it:: 'parses empty: foobar()', fn:
    parse_expr(`foobar()`) eq snapshot


  it:: 'parses single arg: foobar(1)', fn:
    parse_expr(`foobar(1)`) eq snapshot


  it:: 'parses multiple args: foobar(1, 2)', fn:
    parse_expr(`foobar(1, 2)`) eq snapshot


describe:: 'call() - parsing failures', fn:

  it:: `throws when missing ')'`, fn:
    (
      fn: parse_expr(`foobar(`, 'test.fnk')
    ) to_throw `
      test.fnk:1:7
      1| foobar(
                ^

      Expected ')' but found Symbol(end).`


  it:: `throws when missing ')'`, fn:
    (
      fn: parse_expr(`foobar(1:)`, 'test.fnk')
    ) to_throw `
      test.fnk:1:8
      1| foobar(1:)
                 ^

      Expected ')' but found ':'.`


describe:: 'call:: ...', fn:

  it:: 'parses single line', fn:
    parse_expr(`foobar:: spam, ni`) eq snapshot


  it:: 'parses multiline', fn:
    parse_expr(`
      foo::
        fn bar:
          spam
        ni
        shrub
    `) eq snapshot

  it:: 'parses multiline with comma (TODO)', fn:
    parse_expr(`
      foo::
        bar,
        ni,
        shrub
    `) eq snapshot


describe:: 'pipe foo: ...', fn:

  it:: 'pipes', fn:
    parse_expr(`
      pipe foo:
        bar(shrub)
        ni
    `) eq snapshot

