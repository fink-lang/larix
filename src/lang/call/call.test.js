import {parse_expr} from '../..';
import {strip_block} from '../../string-utils';


describe('call()', ()=> {

  it('parses empty: foobar()', ()=> {
    expect(
      parse_expr(`foobar()`)
    ).toMatchSnapshot();
  });


  it('parses single arg: foobar(1)', ()=> {
    expect(
      parse_expr(`foobar(1)`)
    ).toMatchSnapshot();
  });


  it('parses multiple args: foobar(1, 2)', ()=> {
    expect(
      parse_expr(`foobar(1, 2)`)
    ).toMatchSnapshot();
  });
});


describe('call() - parsing failures', ()=> {
  it('throws when missing `)`', ()=> {
    expect(
      ()=> parse_expr(`foobar(`, 'test.fnk')
    ).toThrow(strip_block`
      test.fnk:1:7
      1| foobar(
                ^

      Expected ')' but found Symbol(end).
    `);
  });

  it('throws when missing `,`', ()=> {
    expect(
      ()=> parse_expr(`foobar(1:)`, 'test.fnk')
    ).toThrow(strip_block`
      test.fnk:1:8
      1| foobar(1:)
                 ^

      Expected ',' but found ':'.
    `);
  });
});


describe('call:: ...', ()=> {

  it('parses args: foobar:: spam, ni', ()=> {
    expect(
      parse_expr(`foobar:: spam, ni`)
    ).toMatchSnapshot();
  });
});


describe('pipe foo: ...', ()=> {

  it('pipes', ()=> {
    expect(
      parse_expr(strip_block`
        pipe foo:
          bar(shrub)
          ni
      `)
    ).toMatchSnapshot();
  });
});

