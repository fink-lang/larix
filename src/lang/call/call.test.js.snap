// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`call() parses dangling comma: foobar(1, 2,) 1`] = `
call (1:0-1:13)
  other (1:0-1:6) foobar
  other (1:7-1:8) 1
  other (1:10-1:11) 2

`;

exports[`call() parses empty: foobar() 1`] = `
call (1:0-1:8)
  other (1:0-1:6) foobar

`;

exports[`call() parses multiple args: foobar(1, 2) 1`] = `
call (1:0-1:12)
  other (1:0-1:6) foobar
  other (1:7-1:8) 1
  other (1:10-1:11) 2

`;

exports[`call() parses single arg: foobar(1) 1`] = `
call (1:0-1:9)
  other (1:0-1:6) foobar
  other (1:7-1:8) 1

`;

exports[`call:: ... parses args: foobar:: spam, ni 1`] = `
call (1:0-1:15)
  other (1:0-1:6) foobar
  other (1:9-1:13) spam
  other (1:15-1:17) ni

`;

exports[`pipe foo: ... pipes 1`] = `
block pipe (1:0-3:4)
  other (1:5-1:8) foo
  :
  call (2:2-2:12)
    other (2:2-2:5) bar
    other (2:6-2:11) shrub

  other (3:2-3:4) ni
`;
